### PR TITLE
On 697 staff ui azure keygen

### PR DESCRIPTION
**Pull request by martin-hornsby**

### CODE REVIEW

File: src/SFA.DAS.AssessorService.Web/Controllers/ExternalApiController.cs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+            var result = await _apiClient.CreateUser(ukprn, username);
+            return RedirectToAction("Index");
+        }
+
+        [HttpPost]
+        public async Task<IActionResult> RemoveAccess(AzureUser viewModel)
+        {
+            var ukprn = _contextAccessor.HttpContext.User.Claims.First(c => c.Type == "http://schemas.portal.com/ukprn")?.Value;
+            var username = _contextAccessor.HttpContext.User.Claims.First(c => c.Type == "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/upn")?.Value;
```
Comment (davidgouge) [inline src/SFA.DAS.AssessorService.Web/Controllers/ExternalApiController.cs]: See these a lot throughout .Web.  I wonder if there's a middleware we can add to auto inject this stuff into controllers?  Similar to the ukprn / email thing in the External API? 🤔

File: src/SFA.DAS.AssessorService.Application.Api.Client/Azure/AzureApiClient.cs [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+                user = await PostPutRequestWithResponse<CreateAzureUserRequest, AzureUser>(httpRequest, request);
+            }
+
+            if (user.AzureId != null)
+            {
+                var subscription = await SubscribeUserToProduct(user.Id, _productId);
+                user.Subscriptions.Add(subscription);
+            }
+
+            //await _organisationsApiClient.Update(new UpdateOrganisationRequest
```
Comment (jcoxhead) [inline src/SFA.DAS.AssessorService.Application.Api.Client/Azure/AzureApiClient.cs]: Picky point:- Commented out code for me should be assigned to the source control system
might be worthwhile in group discussion.

File: src/SFA.DAS.AssessorService.Application.Api.Client/Azure/AzureApiClient.cs [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+
+            if (user != null)
+            {
+                using (var httpRequest = new HttpRequestMessage(HttpMethod.Delete, $"/users/{userId}?api-version=2017-03-01&deleteSubscriptions=true"))
+                {
+                    httpRequest.Headers.Add("If-Match", "*");
+                    await Delete(httpRequest);
+                }
+
+                //var organisation = await _organisationsApiClient.Get(user.Note);
```
Comment (jcoxhead) [inline src/SFA.DAS.AssessorService.Application.Api.Client/Azure/AzureApiClient.cs]: Same as previous  comment ...

File: src/SFA.DAS.AssessorService.Application.Api.Client/Azure/AzureApiClient.cs [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+                //    EndPointAssessorOrganisationId = organisation.EndPointAssessorOrganisationId,
+                //    EndPointAssessorUkprn = organisation.EndPointAssessorUkprn,
+                //    ApiEnabled = false,
+                //    ApiUser = null
+                //});
+            }
+
+        }
+
+        public async Task<IEnumerable<AzureProduct>> ListProducts()
```
Comment (jcoxhead) [inline src/SFA.DAS.AssessorService.Application.Api.Client/Azure/AzureApiClient.cs]: Not sure if we need localisation with resource strings? for this project.

File: src/SFA.DAS.AssessorService.Application.Api.Client/Azure/AzureApiClientBase.cs [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Serialization;
+using SFA.DAS.AssessorService.Application.Api.Client.Clients;
+using SFA.DAS.AssessorService.Application.Api.Client.Exceptions;
+
+namespace SFA.DAS.AssessorService.Application.Api.Client.Azure
+{
+    public abstract class AzureApiClientBase : IDisposable
```
Comment (jcoxhead) [inline src/SFA.DAS.AssessorService.Application.Api.Client/Azure/AzureApiClientBase.cs]: We seem to have the same class used in a few projects now - maybe we should rationalise this?

File: src/SFA.DAS.AssessorService.Web.Staff/Controllers/ExternalApiController.cs [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+using System.Threading.Tasks;
+
+namespace SFA.DAS.AssessorService.Web.Staff.Controllers
+{
+    [Authorize]
+    public class ExternalApiController : Controller
+    {
+        private readonly ILogger<ExternalApiController> _logger;
+        private readonly IAzureApiClient _apiClient;
+        private readonly ISessionService _sessionService;
```
Comment (jcoxhead) [inline src/SFA.DAS.AssessorService.Web.Staff/Controllers/ExternalApiController.cs]: _sessionService not used?

File: src/SFA.DAS.AssessorService.Web/Controllers/ExternalApiController.cs [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+namespace SFA.DAS.AssessorService.Web.Controllers
+{
+    [Authorize]
+    public class ExternalApiController : Controller
+    {
+        private readonly ILogger<ExternalApiController> _logger;
+        private readonly IHttpContextAccessor _contextAccessor;
+        private readonly IAzureApiClient _apiClient;
+        private readonly ISessionService _sessionService;
```
Comment (jcoxhead) [inline src/SFA.DAS.AssessorService.Web/Controllers/ExternalApiController.cs]: Unused private variables

File: src/SFA.DAS.AssessorService.Web/Controllers/ExternalApiController.cs [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+            return RedirectToAction("Index");
+        }
+
+        [HttpPost]
+        public async Task<IActionResult> RemoveAccess(AzureUser viewModel)
+        {
+            var ukprn = _contextAccessor.HttpContext.User.Claims.First(c => c.Type == "http://schemas.portal.com/ukprn")?.Value;
+            var username = _contextAccessor.HttpContext.User.Claims.First(c => c.Type == "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/upn")?.Value;
+
+            if (!string.Equals(viewModel.State, "blocked"))
```
Comment (jcoxhead) [inline src/SFA.DAS.AssessorService.Web/Controllers/ExternalApiController.cs]: Use of magic strings maybe we should use const

Comment (martin-hornsby) [reply src/SFA.DAS.AssessorService.Web/Controllers/ExternalApiController.cs]: Feels a bit superfluous in this case. It is an isolated, the value won't ever change and it's self explanatory as to what it's being used for.

File: src/SFA.DAS.AssessorService.Application.Api.Client/Azure/AzureApiClient.cs (hunk 4 above)
Comment (martin-hornsby) [reply src/SFA.DAS.AssessorService.Application.Api.Client/Azure/AzureApiClient.cs]: Localisation hasn't been spec'ed as of yet. @davidgouge could you let me know if you want this.

File: src/SFA.DAS.AssessorService.Web/Controllers/ExternalApiController.cs (hunk 1 above)
Comment (martin-hornsby) [reply src/SFA.DAS.AssessorService.Web/Controllers/ExternalApiController.cs]: I tried to add this but for some reason the middleware couldn't load up the HttpContext correctly

**Pull request merged**